### ISSUE TITLE
Getting malformed JSON

### ISSUE THREAD

**Issue by marcbr8** [sachin-handiekar/jInstagram#184]
Since late October, when calling Instagram.getMediaComments or Instagram.getRecentMediaNextPage, I am running many many times into
```
org.jinstagram.exceptions.InstagramException: Error parsing json to object type org.jinstagram.entity.comments.MediaCommentsFeed
        at org.jinstagram.InstagramBase.createObjectFromResponse(InstagramBase.java:1010) ~[jInstagram-1.1.8.jar:na]
        at org.jinstagram.InstagramBase.createInstagramObject(InstagramBase.java:848) ~[jInstagram-1.1.8.jar:na]
        at org.jinstagram.InstagramBase.createInstagramObject(InstagramBase.java:872) ~[jInstagram-1.1.8.jar:na]
        at org.jinstagram.InstagramBase.getMediaComments(InstagramBase.java:500) ~[jInstagram-1.1.8.jar:na]
```
When logging the JSON that comes from jInstagram, it looks like it's being chopped off.

This is happening hundreds of times per day, although it's difficult to reproduce since one call to one of those methods with the exact same input will sometimes make the exception raise and sometimes will work.

**Comment by abargnesi** [comment]
@marcbr8 I have seen these malformed JSON errors when instagram resets the socket (i.e. connection reset). A connection reset would occur and the response body would be incomplete thus leading to JSON errors.

It seems to be happening to others as well: huttarichard/instagram-private-api#40

Do you see similar connection reset issues? Is there more to that stacktrace?

**Comment by marcbr8** [comment]
Thanks for the fast reply @abargnesi
I am not seeing any errors logged related to a connection/socket reset. More to the stacktrace:
```
org.jinstagram.exceptions.InstagramException: Error parsing json to object type org.jinstagram.entity.comments.MediaCommentsFeed
        at org.jinstagram.InstagramBase.createObjectFromResponse(InstagramBase.java:1010) ~[jInstagram-1.1.8.jar:na]
        at org.jinstagram.InstagramBase.createInstagramObject(InstagramBase.java:848) ~[jInstagram-1.1.8.jar:na]
        at org.jinstagram.InstagramBase.createInstagramObject(InstagramBase.java:872) ~[jInstagram-1.1.8.jar:na]
        at org.jinstagram.InstagramBase.getMediaComments(InstagramBase.java:500) ~[jInstagram-1.1.8.jar:na]
        ...
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) ~[na:1.8.0_101]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) ~[na:1.8.0_101]
        at java.lang.Thread.run(Thread.java:745) ~[na:1.8.0_101]
        Caused by: com.google.gson.JsonSyntaxException: com.google.gson.stream.MalformedJsonException: Unterminated string at line 1 column 27772 path $.data[87].from.profile_picture
        at com.google.gson.Gson.fromJson(Gson.java:894) 
        at com.google.gson.Gson.fromJson(Gson.java:844) 
        at com.google.gson.Gson.fromJson(Gson.java:793) 
        at com.google.gson.Gson.fromJson(Gson.java:765) 
        at org.jinstagram.InstagramBase.createObjectFromResponse(InstagramBase.java:1005) ~[jInstagram-1.1.8.jar:na]
```

**Comment by abargnesi** [comment]
Ok, thanks!

In this case I think you are experiencing a closed socket before transfer completed. This could lead to a truncated response that does not throw an exception.

Take a look at [this gist](https://gist.github.com/abargnesi/e76f33b7b781febb97d16377d8eca1ff) and see if you can reproduce it using HTTP 1.1.

Interestingly, in my tests using HTTP2, I have not seen been able to reproduce the connection reset or truncated response bodies. Instagram supports HTTP2.

**Comment by marcbr8** [comment]
That sounds promising. Any way to configure in the code to use HTTP2?

**Comment by abargnesi** [comment]
Unfortunately, that is not straightforward. This library using Java networking doesn't support HTTP2 (coming in Java 9). We would probably have to use a network library like [netty](https://github.com/netty/netty) to send HTTP2 requests.

**Comment by marcbr8** [comment]
I see. What about until then releasing a version that allows a retry?

**Comment by abargnesi** [comment]
Work was done on this branch https://github.com/abargnesi/jInstagram/tree/retry-on-invalid-json-response-body but I have not sent a PR upstream.

**Comment by marcbr8** [comment]
Sorry for the late reply. I'm testing it now. My first question would be: Any reason to have the retry feature set to false per default?

**Comment by marcbr8** [comment]
Working well! Have gotten a couple malformed jsons, which the retries fixed.

**Comment by abargnesi** [comment]
Glad to hear it works.

`isRetryOnStreamFailure()` defaults to disabled since the retry would consume an additional API call. Additionally a developer may have already provided this type of error handling.

I'll fix up this work and submit a PR.

**Comment by marcbr8** [comment]
Great, let me know if I can help somehow

**Comment by marcbr8** [comment]
Hi @abargnesi I was wondering when you think you'd have time to submit the PR with that fix? If there's anything I can do (more testing, unit tests, refactoring, etc) I would be happy to help. Thanks!

**Comment by tdeekens** [comment]
Any news regarding a PR for the fix? If we can support you in any way let us know!

**Comment by abargnesi** [comment]
This was addressed in #186 and #190.
